src/clients/c++/experimental_api_v2/library/grpc_client.h

```diff
@@ -0,0 +1,140 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+/// \file
+
+#include "src/clients/c++/experimental_api_v2/library/utils.h"
+#include "src/core/constants.h"
+#include "src/core/grpc_service_v2.grpc.pb.h"
+#include "src/core/model_config.pb.h"
+
+namespace nvidia { namespace inferenceserver { namespace client {
+
+//==============================================================================
+/// An InferenceServerGrpcClient object is used to perform any kind of
+/// communication with the InferenceServer using gRPC protocol.
+///
+/// \code
+///   std::unique_ptr<InferenceServerGrpcClient> client;
+///   InferenceServerGrpcClient::Create(&client, "localhost:8001");
+///   std::map<std::string, std::string> headers;
+///   bool live;
+///   client->IsServerLive(&live, headers);
+///   ...
+///   ...
+/// \endcode
+///
+class InferenceServerGrpcClient {
+ public:
+  /// Create a client that can be used to communicate with the server.
+  /// \param client Returns a new InferenceServerGrpcClient object.
+  /// \param server_url The inference server name and port.
+  /// \param verbose If true generate verbose output when contacting
+  /// the inference server.
+  /// \return Error object indicating success or failure.
+  static Error Create(
+      std::unique_ptr<InferenceServerGrpcClient>* client,
+      const std::string& server_url, bool verbose = false);
+
+  /// Contact the inference server and get its liveness.
+  /// \param live Returns whether the server is live or not.
+  /// \param headers Map specifying additional HTTP headers to include in the
+  /// metadata of gRPC request.
+  /// \return Error object indicating success or failure of the request.
+  Error IsServerLive(bool* live, std::map<std::string, std::string>& headers);
+
+  /// Contact the inference server and get its readiness.
+  /// \param ready Returns whether the server is ready or not.
+  /// \param headers Map specifying additional HTTP headers to include in the
+  /// metadata of gRPC request.
+  /// \return Error object indicating success or failure of the request.
+  Error IsServerReady(bool* ready, std::map<std::string, std::string>& headers);
+
+  /// Contact the inference server and get the readiness of specified model.
+  /// \param ready Returns whether the specified model is ready or not.
+  /// \param model_name The name of the model to check for readiness.
+  /// \param headers Map specifying additional HTTP headers to include in the
+  /// metadata of gRPC request.
+  /// \param model_version The version of the model to check for readiness.
+  /// The default value is an empty string which means then the server will
+  /// choose a versionbased on the model and internal policy.
+  /// \return Error object indicating success or failure of the request.
+  Error IsModelReady(
+      bool* ready, const std::string& model_name,
+      std::map<std::string, std::string>& headers,
```

Comment:
Following python APIs it would be better to have model_version after model_name, and default it to empty string. Also, headers is going to be rarely used to forcing create of an empty map to call the API is awkward. Likely want to have 2 IsModelReady functions, one that takes headers and one that doesn't. Similarly for other functions.

Reply:
I have made headers optional too. As the headers would be rarely used, I think we can have single signature. The only thing is that in order to specify headers, user must specify model_version too. Let me know what you think.